### PR TITLE
Remove trait bounds from generated program type's declaration

### PR DESCRIPTION
**Pull request by regexident**
Fixes https://github.com/s-arash/ascent/issues/23

### CODE REVIEW

**Comment by regexident** [comment]
Following up on https://github.com/s-arash/ascent/issues/23#issuecomment-1858083233 I've amended this PR to support the following (optional) syntax for specifying diverging impl trait bounds:

## Non-diverging impl trait bounds

If only a generic `struct` is specified, then the macro will behave as is, using the provided type generics and bounds for both, the type's declaration, as well as its impls:

The following …

```rust
ascent!{
    struct AscentProgram<T> where T: Clone + Eq + Hash;

    // ...
}
```

… expands to something along the lines of this:

```rust
struct AscentProgram<T>
where
   T: Clone + Hash + Eq
{
   // ...
}

impl<T> AscentProgram<T>
where
   T: Clone + Hash + Eq,
{
   // ...
}

impl<T> Default for AscentProgram<T>
where
   T: Clone + Hash + Eq
{
   // ...
}
```

## Diverging impl trait bounds

If both, a generic `struct`, as well as an `impl …` are specified, then the macro use the provided type generics and bounds for the type's declaration, as well as the `impl Default`, while using the provided impl generics and bounds for the type's impls:

The following …

```rust
ascent!{
    struct AscentProgram<T> where T: Clone;
    impl<T> AscentProgram<T> where T: Clone + Eq + Hash;
    // ...
}
```

… expands to something along the lines of this:

```rust
struct AscentProgram<T>
where
    T: Clone
{
   // ...
}

impl<T> AscentProgram<T>
where
   T: Clone + Hash + Eq,
{
   // ...
}

impl<T> Default for AscentProgram<T>
where
    T: Clone
{
   // ...
}
```

## Impl-only trait bounds

Similarly the following …

```rust
ascent!{
    struct AscentProgram<T>;
    impl<T> AscentProgram<T> where T: Clone + Eq + Hash;
    // ...
}
```

… expands to something along the lines of this:

```rust
struct AscentProgram<T> {
   // ...
}

impl<T> AscentProgram<T>
where
   T: Clone + Hash + Eq,
{
   // ...
}

impl<T> Default for AscentProgram<T> {
   // ...
}
```

File: ascent_macro/src/ascent_codegen.rs [hunk 1]
```diff
@@ -226,13 +240,13 @@ pub(crate) fn compile_mir(mir: &AscentMir, is_ascent_run: bool) -> proc_macro2::
    let rule_time_fields_defaults = if mir.config.include_rule_times {rule_time_fields_defaults} else {vec![]};
    let res = quote! {
       #(#struct_attrs)*
-      #vis struct #struct_name #generics {
+      #vis struct #struct_name #ty_impl_generics #ty_where_clause {
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:243]:
```suggestion
      #vis struct #struct_name #ty_generics {
```

File: ascent_macro/src/ascent_codegen.rs [hunk 2]
```diff
@@ -226,13 +240,13 @@ pub(crate) fn compile_mir(mir: &AscentMir, is_ascent_run: bool) -> proc_macro2::
...
-      #vis struct #struct_name #generics {
+      #vis struct #struct_name #ty_impl_generics #ty_where_clause {
          #(#relation_fields)*
          #(#scc_time_fields)*
          #(#rule_time_fields)*
          pub update_time_nanos: std::sync::atomic::AtomicU64,
       }
-      impl #impl_generics #struct_name #ty_generics #where_clause {
+      impl #impl_ty_generics #struct_name #impl_generics #impl_where_clause {
          #run_func
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:250]:
```suggestion
        impl #impl_impl_generics #struct_name #impl_ty_generics #impl_where_clause {
```

File: ascent_macro/src/ascent_codegen.rs [hunk 3]
```diff
@@ -257,7 +271,7 @@ pub(crate) fn compile_mir(mir: &AscentMir, is_ascent_run: bool) -> proc_macro2::
             #scc_times_summary_body
          }
       }
-      impl #impl_generics Default for #struct_name #ty_generics #where_clause {
+      impl #ty_ty_generics Default for #struct_name #ty_ty_generics #ty_where_clause {
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:274]:
```suggestion
        impl #ty_impl_generics Default for #struct_name #ty_ty_generics #ty_where_clause {
```

**Review by s-arash — COMMENTED** [review]
We should also disallow custom `impl` signatures on `ascent_run` and `ascent_run_par`. This could be done at the start of the `compile_mir` function.

**Comment by s-arash** [comment]
Hi @regexident, and thanks again for the PR!
Do you have time to fix the remaining errors in this PR? If not, let me know and I can take it from here.

**Comment by regexident** [comment]
I'll try to give it a go tomorrow or the day after.

**Comment by regexident** [comment]
@s-arash Looks like the [experimental "attrs on relations"](https://github.com/s-arash/ascent/commit/df81b47f95461e9f9c3981f2911d140f3c479535) in `AscentProgram` got broken my changes of the `Signatures` type.

The handling of attritibute parsing within `impl Parse for AscentProgram { … }` looks a bit shady to me (leaky abstraction) since any changes within `Signatures` should not break the parsing of the outer `AscentProgram`.

I'm not sure how to best address this. Feel free to take it from here, if you do.

File: ascent_macro/src/ascent_codegen.rs [hunk 4]
```diff
@@ -212,11 +212,29 @@ pub(crate) fn compile_mir(mir: &AscentMir, is_ascent_run: bool) -> proc_macro2::
...
    let summary = mir_summary(mir);
-   let (impl_generics, ty_generics, where_clause) = mir.declaration.generics.split_for_impl();
-   let vis = &mir.declaration.visibility;
-   let struct_name = &mir.declaration.ident;
-   let generics = &mir.declaration.generics;
-   let struct_attrs = &mir.declaration.attrs;
+   let ty_signature = &mir.signatures.declaration;
+   let impl_signature = &mir.signatures.implementation;
+   
+   let (ty_impl_generics, ty_ty_generics, ty_where_clause) = ty_signature.generics.split_for_impl();
```
Comment (s-arash) [inline ascent_macro/src/ascent_codegen.rs:218]: Could you add a function `split_for_impl()` to `Signatures` and move this logic there?
Also please turn panics into compile errors.

**Review by s-arash — CHANGES_REQUESTED** [review]

**Comment by s-arash** [comment]
> The handling of attritibute parsing within impl Parse for AscentProgram { … } looks a bit shady to me (leaky abstraction) since any changes within Signatures should not break the parsing of the outer AscentProgram.

File: ascent_macro/src/ascent_codegen.rs (hunk 4 above)
Comment (regexident) [reply ascent_macro/src/ascent_codegen.rs]: Done.

**Review by s-arash — APPROVED** [review]

**Pull request merged**